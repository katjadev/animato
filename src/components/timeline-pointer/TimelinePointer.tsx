import { FC, useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { useTranslations } from 'next-intl'
import { ScrollPosition } from '@animato/types'
import { MAX_DURATION, REM_TO_PX_COEFFICIENT } from '@animato/constants';
import styles from './TimelinePointer.module.css'

interface TimelinePointerProps {
  currentTime: number;
  currentPosition: number;
  scrollPosition: ScrollPosition;
  markSize: number;
  timelineWidth: number;
  onChangePosition: (position: number) => void;
}

const TimelinePointer: FC<TimelinePointerProps> = ({
  currentTime,
  currentPosition,
  scrollPosition,
  markSize,
  timelineWidth,
  onChangePosition, 
}) => {
  const t = useTranslations('project.timeline')
  const pointerRef = useRef<HTMLDivElement>(null)
  const [isDragging, setIsDragging] = useState(false)

  const initialPosition = useMemo(() => pointerRef.current?.getBoundingClientRect().x || 0, [pointerRef.current])

  const handleMouseDown = () => setIsDragging(true)

  const handleMouseUp = useCallback(() => {
    if (isDragging) {
      setIsDragging(false)
    }
  }, [isDragging])

  const handleMouseMove = useCallback((event: MouseEvent) => {
    event.preventDefault()
    event.stopPropagation()
    
    if (!isDragging) {
      return
    }

    const newPosition = Math.max(event.clientX - initialPosition + scrollPosition.left, 0)
    onChangePosition(newPosition)
  }, [
    isDragging, 
    initialPosition,
    scrollPosition.left,
    onChangePosition,
  ])

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (event.key === 'ArrowUp' || event.key === 'ArrowRight') {
      event.preventDefault()
      event.stopPropagation()
      const newPosition = Math.min(currentPosition + markSize * REM_TO_PX_COEFFICIENT, timelineWidth)
      onChangePosition(newPosition)
    }

    if (event.key === 'ArrowDown' || event.key === 'ArrowLeft') {
      event.preventDefault()
      event.stopPropagation()
      const newPosition = Math.max(currentPosition - markSize * REM_TO_PX_COEFFICIENT, 0)
      onChangePosition(newPosition)
    }
  }

  useEffect(() => {
    document.addEventListener('mousemove', handleMouseMove)
    document.addEventListener('mouseup', handleMouseUp)
    document.addEventListener('mouseleave', handleMouseUp)

    return () => {
      document.removeEventListener('mousemove', handleMouseMove)
      document.removeEventListener('mouseup', handleMouseUp)
      document.removeEventListener('mouseleave', handleMouseUp)
    }
  }, [
    isDragging, 
    handleMouseMove, 
    handleMouseUp,
  ])

  return (
    <div 
      className={styles.pointer}
      role="slider"
      aria-label={t('pointer-aria-label')}
      aria-valuemin={0}
      aria-valuemax={MAX_DURATION}
      aria-valuenow={currentTime}
      aria-valuetext={t('pointer-aria-valuetext', { time: currentTime })}
      tabIndex={0}
      ref={pointerRef}
      style={{ 
        left: `${currentPosition}px`,
        marginLeft: `-${scrollPosition.left}px`,
      }}
      onMouseDown={handleMouseDown}
      onKeyDown={handleKeyDown}
    />
  )
}

export default TimelinePointer