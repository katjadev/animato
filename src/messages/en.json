{
  "projects": "Projects",
  "logout": "Log out",
  "login": "Log in",
  "home-title": "Bring your designs to life with animated SVGs",
  "home-description": "Animato empowers you to easily create and animate scalable vector graphics, breathing life into your designs and enhancing your web presence with eye-catching visuals that capture your audience's attention.",
  "get-started": "Get Started",
  "explore-demo-project": "Explore a demo project",
  "signup-dialog": {
    "create-account": "Create an account",
    "email": "Email",
    "password": "Password"
  },
  "login-dialog": {
    "log-in": "Log in",
    "email": "Email",
    "password": "Password"
  },
  "project-list": {
    "projects": "Projects",
    "profile": "Profile",
    "new-project": "New project",
    "empty-message": "You don't have any projects yet.",
    "title": "Title",
    "last-modified": "Last modified",
    "actions": "Actions",
    "open": "Open",
    "delete-project": "Delete project"
  },
  "project-header": {
    "profile": "Profile",
    "undo": "Undo",
    "redo": "Redo"
  },
  "project-controls": {
    "restart": "Restart",
    "play": "Play",
    "pause": "Pause",
    "repeat": "Repeat"
  },
  "project": {
    "elements": "Elements",
    "timeline": {
      "pointer-aria-label": "Current time",
      "pointer-aria-valuetext": "{time} milliseconds"
    }
  },
  "main-menu": {
    "import-svg": "Import SVG",
    "export": "Export",
    "back-to-projects": "Back to projects"
  }
}