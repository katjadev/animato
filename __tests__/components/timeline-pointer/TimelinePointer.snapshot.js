import { render } from '@testing-library/react'
import { NextIntlProvider } from 'next-intl'
import messages from '../../../src/messages/en.json'
import TimelinePointer from '../../../src/components/timeline-pointer/TimelinePointer'

it('renders correctly', () => {
  const useRouter = jest.spyOn(require('next/router'), 'useRouter')
  const locale = 'en'
  useRouter.mockImplementationOnce(() => ({
    query: { locale: locale },
  }))

  const { container } = render(
    <NextIntlProvider messages={messages} locale={locale}>
      <TimelinePointer
        currentTime={0}
        currentPosition={0}
        scrollPosition={{ left: 0, top: 0 }}
        onChangePosition={() => {}}
      />
    </NextIntlProvider>
  )
  expect(container).toMatchSnapshot()
})
