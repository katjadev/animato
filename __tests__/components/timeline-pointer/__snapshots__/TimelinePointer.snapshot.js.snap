// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    aria-label="Current time"
    aria-valuemax="300000"
    aria-valuemin="0"
    aria-valuenow="0"
    aria-valuetext="0 milliseconds"
    class="pointer"
    role="slider"
    style="left: 8px; margin-left: -0px;"
    tabindex="0"
  />
</div>
`;
